From 2fe8d723ec0db70cf719aec1bf4bf67726e06c92 Mon Sep 17 00:00:00 2001
From: Yevgeny Rouban <yevgeny.y.rouban@intel.com>
Date: Tue, 25 Jun 2013 11:42:49 +0700
Subject: Dalvik: skip reporting zero-sized code blocks to VTune collector

BZ: 116505

The VTune JIT collector (libittnotify.so) reports an error if a method
is sent with size=0. It is OK if some methods become empty (like
common_invokeMethodNoRange_noJmp or common_invokeMethodRange_noJmp).
In this case it makes no sense to send empty methods to VTune because
no address will be mapped to those methods.

So the patch just skips reporting zero-sized methods to the VTune collector.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I5a6389a331fe87f1a0ad30c0e1843e03556568ad
Orig-MCG-Change-Id: If28ea3b328d21ec18feec4ba3a286269297c3cf6
Signed-off-by: Yevgeny Rouban <yevgeny.y.rouban@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/Lower.cpp |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/codegen/x86/Lower.cpp b/vm/compiler/codegen/x86/Lower.cpp
index 324043d..a84c98a 100644
--- a/vm/compiler/codegen/x86/Lower.cpp
+++ b/vm/compiler/codegen/x86/Lower.cpp
@@ -1161,6 +1161,10 @@ int op_nop(const MIR * mir) {
  * Send the label information (size, start_address and name) to VTune
  */
 void sendLabelInfoToVTune(int startStreamPtr, int endStreamPtr, const char* labelName) {
+    if (endStreamPtr == startStreamPtr) {
+        return;
+    }
+
     iJIT_Method_Load jitMethod;
     memset(&jitMethod, 0, sizeof(iJIT_Method_Load));
     jitMethod.method_id = iJIT_GetNewMethodID();
-- 
1.7.4.1

